Makefile

```diff
@@ -14,7 +14,55 @@ build:
 	'$(VIRTUALENV_DIR)/bin/pip' install -r ./requirements.txt
 	'$(VIRTUALENV_DIR)/bin/python' setup.py install
 
+# Create a image with docker build & docker run.
+docker-build:
+	# e.g., docker build -t  <image_name> /path/to/Dockerfile/
+	# -t flag - to tag the image with a name so Docker recognizes this name.
+	docker build -t migrate2to3:v1 .
+
+docker-run: docker-build
+	# Runs docker-build part before docker-run part, i.e., docker-run DEPENDS on docker-build.
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 bash
+
+# Create 2 virtualenvs & test if each virtualenv with each python is running correctly.
+docker-py27env:
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 virtualenv py27env -p /usr/bin/python
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 py27env/bin/pip2 install -r requirements.txt
+
+docker-py36env:
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 virtualenv py36env -p /usr/bin/python3
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 py36env/bin/pip3 install -r requirements.txt
+
+test_py2.7: docker-py27env
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py27env/bin/python ./tablesnap --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap -e USER=root migrate2to3:v1 ./py27env/bin/python ./tableslurp --help
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py27env/bin/python ./tablechop --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py27env/bin/python ./tablesnap-sunset --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py27env/bin/python ./tablesnap-sync-manifest --help
+
+test_py3.6: docker-py36env # Note: Expect this to FAIL until Python migration changes done when syntax errors corrected.
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py36env/bin/python ./tablesnap --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap -e USER=root migrate2to3:v1 ./py36env/bin/python ./tableslurp --help
+	docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py36env/bin/python ./tablechop --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py36env/bin/python ./tablesnap-sunset --help
+	# docker run -it --mount type=bind,source=$(PWD),target=/tablesnap migrate2to3:v1 ./py36env/bin/python ./tablesnap-sync-manifest --help
+
+DIRECTORY=/tmp/tablesnap-test
+TIMEOUT_A=5
+TIMEOUT_B=10
+AWS_ACC_KEY_ID := $(shell aws configure get aws_access_key_id)
+AWS_SEC_ACC_KEY := $(shell aws configure get aws_secret_access_key)
+BUCKET=klew-tablesnap-pymigration
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+docker-test-tablesnap: docker-build
```

Comment:
This `docker-test-tablesnap` target really needs not just `docker-build` but either `docker-py27env` or `docker-py36env` targets, since you need at least one of the installed Python environments in order to run the test script.

However, since we've got particular versions of Python that won't really change often, I suggest you create the Python 2.7 and Python 3.6 environments and install all of the required Python packages as part of `docker-build`. That way you can get rid of the `docker-py27env` and `docker-py36env` targets up above. And you don't need to change the dependencies for `docker-test-tablesnap`.

And despite having talked you into using the "bind" mount before, I think we might simplify things if we get rid of it.
You can use the `COPY` command in your `Dockerfile` to copy `tablesnap` and its companion programs to where you need them. This will get rid of the `py27env` and `py36env` directories you had to put in `.gitignore`, and also get rid of the long `--mount` option you have to give to each of the `docker run` commands.

So here's what you'd need to do in your `Dockerfile`.

After the `WORKDIR` command at the bottom, add the commands

```
RUN virtualenv py27env -p /usr/bin/python
RUN virtualenv py36env -p /usr/bin/python3

COPY requirements.txt .

RUN py27env/bin/pip2 install -r requirements.txt
RUN py36env/bin/pip3 install -r requirements.txt

COPY tablechop tablesnap ...(other programs) .
COPY test_tablesnap.sh .
```

Then you should be able to remove the `docker-py27env` and `docker-py36env` targets entirely, and also remove the long `--mount` option

```
--mount type=bind,source=$(PWD),target=/tablesnap
```

from all the `docker run` commands.

Reply:
Once you have done this, there are a couple of other things to tweak but I think it will be easier to explain if you do this change first.